### ISSUE TITLE
java.sql.SQLException: Specified key was too long; max key length is 767 bytes

### ISSUE THREAD

**Issue by ae-g-i-s** [airsonic/airsonic#725]
## Problem description

*TL; DR: Airsonic's varchar size is too large for primary keys in MySQL < 5.7.9/MariaDB < 10.2.4*.

MariaDB 10.1.31 and MySQL 5.7 as well as most other versions of the MySQL/MariaDB universe have a limitation of 767 bytes for indices in InnoDB (see [a related StackOverflow thread](https://stackoverflow.com/questions/1814532/1071-specified-key-was-too-long-max-key-length-is-767-bytes) and [the MySQL docs](https://dev.mysql.com/doc/refman/5.7/en/create-index.html) ).

The exact allowed length depends on the charset; with utf-8, it seems to be 255 characters. It is the aforementioned limit divided by the largest possible byte size of a single character - in utf-8, that's three bytes.

`WEB-INF/classes/liquibase/legacy/schema25.xml` says:
```xml
<column name="username" type="${varchar_type}">
  <constraints primaryKey="true" />
</column>
```

`WEB-INF/classes/liquibase/db-changelog.xml` says:
```xml
<property name="varchar_type" dbms="mysql" value="varchar(${mysqlVarcharLimit})" />
<property name="varchar_type" value="varchar" />
```

`WEB-INF/classes/applicationContext-db.xml` defines that as:
```xml
<entry key="mysqlVarcharLimit" value="${DatabaseMysqlMaxlength:512}" />
```

This leads to the following error when trying to start Airsonic:
```
liquibase.exception.DatabaseException: (conn=5) Specified key was too long; max key length is 767 bytes [Failed SQL: CREATE TABLE airsonic.user (username VARCHAR(512) NOT NULL, password VARCHAR(512) NOT NULL, CONSTRAINT PK_USER PRIMARY KEY (username))]
```

### Steps to reproduce

Modify `airsonic.properties` to use embedded JDBC and start Tomcat.

## System information

 * **Airsonic version**: v10.1.1
 * **Database**: MariaDB 10.1.31

**Comment by ae-g-i-s** [comment]
I just saw that there is a setting for the varchar size in the admin panel. However, this is not documented on https://airsonic.github.io/docs/database/; after manually overwriting the `mysqlVarcharLimit` entry in `WEB-INF/classes/applicationContext-db.xml` and setting it to 255, everything seems to be working.
I suppose the sane way to handle this is either
* set the default MySQL varchar limit to 255 (*not 256*, that's already too large) instead of 512 **or**
* use a smaller varchar limit for all columns used in indices.

**Comment by hyptechdev2015** [comment]
thanks for the info

**Comment by kaysond** [comment]
So I'm trying to get this up and running, and I have MariaDB 10.3.7. Unfortunately, airsonic still crashes with `Specified key was too long; max key length is 3072 bytes [Failed SQL: ALTER TABLE airsonic.user_rating ADD PRIMARY KEY (username, path)]`. Presumably this is because the index is on two `varchar(512)`'s? Is there a way to fix it besides modifying the source myself?

**Comment by kaysond** [comment]
So the issue in my case was that the default collation was `utf8mb4_general_ci` which uses the full 4 bytes per character. Changing the default database collation to `utf8_general_ci` (or `utf8_unicode_ci`) fixed the problem.

**Comment by Sephtex** [comment]
I got this issue solved with the following configuration for mariadb 10.1.32 and Airsonic 10.1.2:
`
    - { option: innodb_file_format,              value: Barracuda }
    - { option: innodb_large_prefix,             value: 'ON' }
    - { option: innodb_default_row_format,       value: DYNAMIC }
`
And database charset/collate:
´
    db_character_set: utf8
    db_collate: utf8_general_ci
´

**Comment by jvoisin** [comment]
@Sephtex can you (pretty) please put this down in a PR in the [documentation's repository]( https://github.com/airsonic/documentation) ?

**Comment by Sephtex** [comment]
@jvoisin I created the following PR https://github.com/airsonic/documentation/pull/66

**Comment by jvoisin** [comment]
Can we mark this as solved?

**Comment by kaysond** [comment]
This is somewhat old now, but my original problem was with the default settings of mariadb 10.3.7 on Ubuntu 18.04.2, which is >10.2. Unless airsonic explicitly sets the collation on creating the db now, I'd say there's still a problem...

**Comment by hasamba** [comment]
im getting the same message,
version 10.6.1
i just upgrade my mysql to 5.6.46 for this but this is what im getting (copied from terminal):
      
org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1131) ~[spring-beans-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1059) ~[spring-beans-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:583) ~[spring-beans-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
        ... 54 common frames omitted
Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:liquibase/legacy/schema25.xml::schema25_004::muff1nman:
     Reason: liquibase.exception.DatabaseException: Specified key was too long; max key length is 767 bytes [Failed SQL: (1071) CREATE TABLE airsonic.user (username VARCHAR(512) NOT NULL, password VARCHAR(512) NOT NULL, CONSTRAINT PK_USER PRIMARY KEY (username))]
        at liquibase.changelog.ChangeSet.execute(ChangeSet.java:646) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:53) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:83) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.Liquibase.update(Liquibase.java:202) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.Liquibase.update(Liquibase.java:179) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.integration.spring.SpringLiquibase.performUpdate(SpringLiquibase.java:366) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:314) ~[liquibase-core-3.8.7.jar!/:na]
        at org.airsonic.player.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:25) ~[classes!/:10.6.1-RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1688) ~[spring-beans-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1626) ~[spring-beans-4.3.25.RELEASE.jar!/:4.3.25.RELEASE]
        ... 66 common frames omitted
Caused by: liquibase.exception.DatabaseException: Specified key was too long; max key length is 767 bytes [Failed SQL: (1071) CREATE TABLE airsonic.user (username VARCHAR(512) NOT NULL, password VARCHAR(512) NOT NULL, CONSTRAINT PK_USER PRIMARY KEY (username))]
        at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:402) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:59) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:131) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.database.AbstractJdbcDatabase.execute(AbstractJdbcDatabase.java:1276) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1258) ~[liquibase-core-3.8.7.jar!/:na]
        at liquibase.changelog.ChangeSet.execute(ChangeSet.java:609) ~[liquibase-core-3.8.7.jar!/:na]
        ... 75 common frames omitted
Caused by: java.sql.SQLSyntaxErrorException: Specified key was too long; max key length is 767 bytes
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120) ~[mysql-connector-java-8.0.19.jar!/:8.0.19]
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97) ~[mysql-connector-java-8.0.19.jar!/:8.0.19]
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122) ~[mysql-connector-java-8.0.19.jar!/:8.0.19]
        at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:764) ~[mysql-connector-java-8.0.19.jar!/:8.0.19]
        at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:648) ~[mysql-connector-java-8.0.19.jar!/:8.0.19]
        at org.apache.commons.dbcp2.DelegatingStatement.execute(DelegatingStatement.java:194) ~[commons-dbcp2-2.7.0.jar!/:2.7.0]
        at org.apache.commons.dbcp2.DelegatingStatement.execute(DelegatingStatement.java:194) ~[commons-dbcp2-2.7.0.jar!/:2.7.0]
        at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:398) ~[liquibase-core-3.8.7.jar!/:na]
        ... 80 common frames omitted
```
